onnxscript/function_libs/tools/torch_lib/deduce_type_constraints.py

```diff
@@ -0,0 +1,403 @@
+        return f"Value(name={self.name}, type_constraint={self.type_constraint})"
+
+
+@dataclasses.dataclass
+class OnnxFunctionTypeConstraints:
+    input_type_constraints: Dict[str, Optional[TypeConstraint]]
+    output_type_constraints: Dict[str, Optional[TypeConstraint]]
+    intermediate_type_constraints: Dict[str, Optional[TypeConstraint]]
+
+    def __repr__(self):
+        repr_strs = [
+            "Type Constraints:",
+            "  Inputs: ",
+        ]
+        repr_strs += [
+            f"    {name}: {type_constraint.name}"
+            if type_constraint is not None
+            else f"    {name}: None"
+            for name, type_constraint in self.input_type_constraints.items()
+        ]
+        repr_strs += [
+            "  Outputs: ",
+        ]
+        repr_strs += [
+            f"    {name}: {type_constraint.name}"
+            if type_constraint is not None
+            else f"    {name}: None"
+            for name, type_constraint in self.output_type_constraints.items()
+        ]
+        repr_strs += [
+            "  Type Constraints: ",
+        ]
+        # Trick to get unique type constraints but maintain the order.
+        ordered_unique_type_constraints = {
+            v: None for v in self.input_type_constraints.values()
+        }
+        ordered_unique_type_constraints.update(
+            {v: None for v in self.output_type_constraints.values()}
+        )
+        repr_strs += [
+            f"    {type_constraint.name}: {type_constraint.type_strs}"
+            for type_constraint in ordered_unique_type_constraints.keys()
+            if type_constraint is not None
+        ]
+
+        if self.intermediate_type_constraints:
+            repr_strs += ["  Intermediate Values: "]
+            repr_strs += [
+                f"    {name}: {type_constraint.name}"
+                if type_constraint is not None
+                else f"    {name}: None"
+                for name, type_constraint in self.intermediate_type_constraints.items()
+            ]
+
+            repr_strs += [
+                "  Intermediate Type Constraints: ",
+            ]
+            ordered_unique_type_constraints = {
+                v: None for v in self.intermediate_type_constraints.values()
+            }
+            repr_strs += [
+                f"    {type_constraint.name}: {type_constraint.type_strs}"
+                for type_constraint in ordered_unique_type_constraints.keys()
+                if type_constraint is not None
+            ]
+
+        return "\n".join(repr_strs)
+
+
+class TypeConstraintDeducer:
+    def __init__(self, onnx_function: onnxscript.OnnxFunction):
+        self.onnx_function = onnx_function
+        self.values: Dict[str, Value] = {}
+
+    def type_constraints(self, signature_only: bool = True) -> OnnxFunctionTypeConstraints:
+        """Retrieve deduced type constraints for the ONNX function."""
+        if not self.values:
+            raise ValueError("Must call deduce() first")
```

Comment:
Potentially reorganize the logic to make type_constraints() and deduce() more separate with clear responsibilities.